### PR TITLE
Add aggregation apiserver and move cluster crd to aa

### PR DESCRIPTION
**Pull request by XiShanYongYe-Chang**
**What type of PR is this?**

/kind feature
/kind api-change

**What this PR does / why we need it**:

As discussed in #1077 and the [meeting agenda on 2021-12-07](https://docs.google.com/document/d/1y6YLVC-v7cmVAdbjedoyR5WL0-q45DBRXTvz5_I7bkA/edit#heading=h.pk408gvppe2h), we need to support a feature to access member clusters through karmada.

We addopt aggregation apiserver to connect karmada apiserver, and move cluster crd to aggregation apiserver.

**Which issue(s) this PR fixes**:
Fixes #1077 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Move cluster crd to aggregation apiserver.
```

/cc @RainbowMango @GitHubxsy @mrlihanbo

### CODE REVIEW

**Comment by pigletfly** [comment]
is there any other use of the AA server ?

**Comment by XiShanYongYe-Chang** [comment]
> is there any other use of the AA server ?

Add cluster proxy like this `/apis/cluster.karmada.io/v1alpha1/cluster/{clustername}/proxy/api/v1/nodes`, just as `service/proxy`, `node/proxy`, `pod/proxy`. Current pr is the first step to add AA server.

**Comment by GitHubxsy** [comment]
> is there any other use of the AA server ?

Add a new  component named cluster-aggregated-apiserver to access the member cluster

image

**Comment by kevin-wangzefeng** [comment]
Adding reference to previous issue discussions would be very helpful to people don't know the background.
@XiShanYongYe-Chang

File: artifacts/deploy/karmada-aggregation-apiserver.yaml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            - --tls-cert-file=/etc/kubernetes/pki/karmada.crt
+            - --tls-private-key-file=/etc/kubernetes/pki/karmada.key
+            - --audit-log-path=-
+            - --feature-gates=APIPriorityAndFairness=false
+            - --audit-log-maxage=0
+            - --audit-log-maxbackup=0
+          resources:
+            requests:
+              cpu: 100m
+              memory: 20Mi
```
Comment (Garrybest) [inline artifacts/deploy/karmada-aggregation-apiserver.yaml:52]: 20MiB, is it enough?

File: pkg/apis/cluster/types.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+)
+
+//revive:disable:exported
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Cluster represents the desire state and status of a member cluster.
+type Cluster struct {
```
Comment (Garrybest) [inline pkg/apis/cluster/types.go:15]: Where is the json annotations of this struct? Like `json:"status,omitempty"`?

File: pkg/aggregationapiserver/apiserver.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
+package aggregationapiserver
```
Comment (Garrybest) [inline pkg/aggregationapiserver/apiserver.go:1]: Are we using [`apiserver-runtime`](https://github.com/kubernetes-sigs/apiserver-runtime) to help us build the AA?

**Comment by Garrybest** [comment]
And how do we upgrade from CRD to AA? Do we have a safe plan?

**Comment by XiShanYongYe-Chang** [comment]
> Adding reference to previous issue discussions would be very helpful to people don't know the background. 

I have added the discussion #1077 to the reference.

File: artifacts/deploy/karmada-aggregation-apiserver.yaml (hunk 1 above)
Comment (XiShanYongYe-Chang) [reply artifacts/deploy/karmada-aggregation-apiserver.yaml:52]: It does seem to be too small. 

It's ok during my local testing.

Let me remove it first just as `karmada-apiserver`.

File: pkg/apis/cluster/types.go (hunk 2 above)
Comment (XiShanYongYe-Chang) [reply pkg/apis/cluster/types.go:15]: It is a internal type of `Cluster`, refer to:

https://github.com/kubernetes/kubernetes/blob/489fb9bee3f626b3eeb120a5af89ad8c2b2f1c20/pkg/apis/apps/types.go#L315-L318

File: pkg/aggregationapiserver/apiserver.go (hunk 3 above)
Comment (XiShanYongYe-Chang) [reply pkg/aggregationapiserver/apiserver.go:1]: Without using `apiserver-runtime` to build the AA.  Directly use `k8s apiserver` just as https://github.com/kubernetes/sample-apiserver.

**Comment by XiShanYongYe-Chang** [comment]
> And how do we upgrade from CRD to AA? Do we have a safe plan?

Yes. AA and `karmada-apiserver` are connected to the same etcd,  we can migrate etcd data with a new prefix. The upgrade plan will be provided later.

File: artifacts/deploy/karmada-apiserver.yaml [hunk 4]
```diff
@@ -54,10 +54,10 @@ spec:
             - --service-account-key-file=/etc/kubernetes/pki/karmada.key
             - --service-account-signing-key-file=/etc/kubernetes/pki/karmada.key
             - --service-cluster-ip-range=10.96.0.0/12
-            - --proxy-client-cert-file=/etc/kubernetes/pki/karmada.crt
-            - --proxy-client-key-file=/etc/kubernetes/pki/karmada.key
+            - --proxy-client-cert-file=/etc/kubernetes/pki/front-proxy-client.crt
+            - --proxy-client-key-file=/etc/kubernetes/pki/front-proxy-client.key
```
Comment (RainbowMango) [inline artifacts/deploy/karmada-apiserver.yaml:58]: If we change the parameter:
@prodanlabs we should update `karamda init` right?
@jrkeen we should update `chart` right?

Comment (RainbowMango) [reply artifacts/deploy/karmada-apiserver.yaml:58]: You might be confused about this PR. We'll talk about this PR at tomorrow's community meeting.

Comment (jrkeen) [reply artifacts/deploy/karmada-apiserver.yaml:58]: Yes, the chart should be updated.

File: pkg/apis/cluster/types.go (hunk 2 above)
Comment (RainbowMango) [reply pkg/apis/cluster/types.go:15]: I asked the same question too. :)

File: pkg/util/gclient/gclient.go [hunk 5]
```diff
@@ -21,6 +22,7 @@ var aggregatedScheme = runtime.NewScheme()
 
 func init() {
 	var _ = scheme.AddToScheme(aggregatedScheme)             // add Kubernetes schemes
+	var _ = clusterapis.AddToScheme(aggregatedScheme)        // add cluster schemes
```
[the remark this answers is not in the document]
Comment (XiShanYongYe-Chang) [reply pkg/util/gclient/gclient.go]: After verification, it is not mandatory.

File: artifacts/deploy/karmada-apiserver.yaml (hunk 4 above)
Comment (prodanlabs) [reply artifacts/deploy/karmada-apiserver.yaml:58]: Yes, i follow up

**Review by RainbowMango — COMMENTED** [review]
/lgtm
/approve

**Comment by RainbowMango** [comment]
In order not to block following works, we are going to merge this first.

**Pull request merged**